loopy/statistics.py

```diff
@@ -170,51 +181,201 @@ def map_if_positive(self, expr):  # implemented in FlopCounter
 
     def map_derivative(self, expr):
-        raise NotImplementedError("OpCounter encountered derivative, "
+        raise NotImplementedError("ExpressionOpCounter encountered derivative, "
                                   "map_derivative not implemented.")
         return 0
 
     def map_slice(self, expr):
-        raise NotImplementedError("OpCounter encountered slice, "
+        raise NotImplementedError("ExpressionOpCounter encountered slice, "
                                   "map_slice not implemented.")
         return 0
 
 
-class SubscriptCounter(CombineMapper):
-    def __init__(self, kernel):
-        self.kernel = kernel
+class ExpressionSubscriptCounter(CombineMapper):
+
+    def __init__(self, knl):
+        self.knl = knl
+        from loopy.expression import TypeInferenceMapper
+        self.type_inf = TypeInferenceMapper(knl)
 
     def combine(self, values):
         return sum(values)
 
+    def map_constant(self, expr):
+        return TypeToCountMap()
+
+    map_tagged_variable = map_constant
+    map_variable = map_constant
+    map_call = map_constant
+
     def map_subscript(self, expr):
-        name = expr.aggregate.name
-        arg = self.kernel.arg_dict.get(name)
-        tv = self.kernel.temporary_variables.get(name)
-        if arg is not None:
-            if isinstance(arg, lp.GlobalArg):
-                # It's global memory
-                pass
-        elif tv is not None:
-            if tv.is_local:
-                # It's shared memory
-                pass
-        return 1 + self.rec(expr.index)
+        name = expr.aggregate.name  # name of array
 
-    def map_constant(self, expr):
+        if name in self.knl.arg_dict:
+            array = self.knl.arg_dict[name]
+        else:
+            # this is a temporary variable
+            return self.rec(expr.index)
+
+        if not isinstance(array, lp.GlobalArg):
+            # this array is not in global memory
+            return self.rec(expr.index)
+
+        index = expr.index  # could be tuple or scalar index
+        if not isinstance(index, tuple):
+            index = (index,)
+
+        from loopy.symbolic import get_dependencies
+        from loopy.kernel.data import LocalIndexTag
+        my_inames = get_dependencies(index) & self.knl.all_inames()
+        local_id0 = None
+        local_id_found = False
+        for iname in my_inames:
+            # find local id0
+            tag = self.knl.iname_to_tag.get(iname)
+            if isinstance(tag, LocalIndexTag):
+                local_id_found = True
+                if tag.axis == 0:
+                    local_id0 = iname
+                    break  # there will be only one local_id0
+
+        if not local_id_found:
+            # count as uniform access
+            warnings.warn("ExpressionSubscriptCounter did not find "
+                          "local iname tags in expression:\n %s,\n"
+                          "considering these DRAM accesses uniform." % expr)
```

Comment:
Wouldn't this warn about every single uniform access? There's nothing alarming about not finding any LocalIndexTag to me. Is there a specific reason this is a warning? Warnings to my mind apply when something is almost certainly wrong (e.g. an unsynchronized write race--we can generate code for it, but it's unlikely to do anything meaningful. There shouldn't be a risk of triggering a warning as you're just going about your business though.